### PR TITLE
Fixed #22276 -- Surfaced ValidationError to end-user if management form is invalid

### PR DESCRIPTION
**Pull request by patrys**
This is a new implementation of #3634

### CODE REVIEW

**Comment by patrys** [comment]
@timgraham please review, I've fixed the cosmetic changes and rebased on current master. I'm not sure whether the end user should be told to file a bug. Do you have an idea in mind how error customization would work? Should it become a class attribute?

File: django/forms/formsets.py [hunk 1]
```diff
@@ -297,6 +296,10 @@ def is_valid(self):
         """
         if not self.is_bound:
             return False
+        if not self.management_form.is_valid():
+            error = _('Management form data is missing or has been tampered with.')
```
Comment (mjtamlyn) [inline django/forms/formsets.py]: This should be overridable by `default_error_message` and passing `error_messages` to the formset for consistency with `Field`.

File: django/forms/formsets.py [hunk 2]
```diff
@@ -297,6 +296,10 @@ def is_valid(self):
...
+            self._non_form_errors = self.error_class([error])
```
Comment (mjtamlyn) [inline django/forms/formsets.py]: This should pass a `ValidationError` with a code.

**Comment by mjtamlyn** [comment]
Needs docs, release notes and tests for the explicit output.

As a side note, I'm really not convinced this is a good idea. I don't hold that a missing management form is a user error - it's generally almost always an application error (failing to output the management form to the page). A developer who encounters this error is likely to be laying out the formset by hand - missing the management form. It's reasonable to assume they're also not going to output the `non_form_errors` in this case, making it highly likely that the user sees their forms come back empty and with no error message explaining it. (If the management form is not rendered, its errors won't be either). This is a worse end user experience than a 500.

The only counterargument I can see is it allows an end user to generate a 500 on any Django site with a formset.

**Comment by apollo13** [comment]
@mjtamlyn With bots running around and filling in arbitrary stuff I think not erroring out if the management form is invalid is a good idea.

**Comment by carljm** [comment]
@mjtamlyn I agree with @apollo13. I think your analysis is correct, but I think that not allowing end users to generate 500 errors at will is more important than the potential debugging issues, in this case. I'm in favor of this PR.

**Comment by carljm** [comment]
There are already other existing situations where you will see an empty form re-display with no apparent problems if you've failed to output non-form-errors (or non-field-errors on a form), so this isn't really adding a new debugging problem, it's just making one more small case behave the same as many other existing cases, instead of having it behave differently. "Check the form(set) errors if the form(set) is invalid and you don't know why" is already a required part of the debugging toolbox for anyone working with form(sets); this PR doesn't change that.

**Comment by timgraham** [comment]
Closing due to inactivity.

**Pull request closed without merging**